tests/python/relax/test_dataflow_rewriter.py

```diff
@@ -0,0 +1,1370 @@
+        else:
+            C = R.call_pure_packed(
+                "my_optimized_add_impl", A, B, sinfo_args=R.Tensor([16], "float32")
+            )
+            out = R.call_pure_packed(
+                "my_optimized_add_impl", C, B, sinfo_args=R.Tensor([16], "float32")
+            )
+        return out
+
+    after = Rewriter(before)
+    tvm.ir.assert_structural_equal(expected, after)
+
+
+def test_match_dynamic_shape():
+    """Pattern match/rewrites may be dynamic
+
+    The tuple being replaced does not need to explicitly exist within
+    the updated Relax function.  So long as each element of the tuple
+    pattern matches a Relax expression, the pattern match can apply.
+
+    This rule ensures that pattern-matching is never broken when
+    `CanonicalizeBindings` is applied.
+
+    This test is identical to `test_rewrite_of_explicit_relax_tuple`,
+    except that the function does not contain the round trip of
+    packing `proj_A` and `proj_B` into a tuple, then immediately
+    unpacking them from the tuple.
+
+    """
+
+    @R.rewriter
+    class Rewriter:
+        @R.function
+        def pattern(
+            lhs_A: R.Tensor(["N1", "M"], "float32"),
+            lhs_B: R.Tensor(["N2", "M"], "float32"),
+            rhs: R.Tensor(["M"], "float32"),
+        ):
+            proj_A = R.matmul(lhs_A, rhs)
+            proj_B = R.matmul(lhs_B, rhs)
+            return (proj_A, proj_B)
+
+        @R.function
+        def replacement(
+            lhs_A: R.Tensor(["N1", "M"], "float32"),
+            lhs_B: R.Tensor(["N2", "M"], "float32"),
+            rhs: R.Tensor(["M"], "float32"),
+        ):
+            N1 = T.int64()
+            N2 = T.int64()
+
+            lhs = R.concat([lhs_A, lhs_B])
+            proj_concat = R.matmul(lhs, rhs)
+            proj_A: R.Tensor([N1], "float32") = R.strided_slice(
+                proj_concat, axes=[0], begin=[0], end=[N1]
+            )
+            proj_B: R.Tensor([N2], "float32") = R.strided_slice(
+                proj_concat, axes=[0], begin=[N1], end=[N2 + N1]
+            )
+            return (proj_A, proj_B)
+
+    @R.function(private=True)
+    def before(
+        state: R.Tensor([16], "float32"),
+        A: R.Tensor([16, 16], "float32"),
+        B: R.Tensor([16, 16], "float32"),
+    ):
+        proj_A = R.matmul(A, state)
+        proj_B = R.matmul(B, state)
+        out = proj_A + proj_B
+        return out
+
+    @R.function(private=True)
+    def expected(
+        state: R.Tensor([16], "float32"),
+        A: R.Tensor([16, 16], "float32"),
+        B: R.Tensor([16, 16], "float32"),
+    ):
+        concat_AB = R.concat([A, B])
+        proj_concat = R.matmul(concat_AB, state)
+        proj_A = R.strided_slice(proj_concat, axes=[0], begin=[0], end=[16])
+        proj_B = R.strided_slice(proj_concat, axes=[0], begin=[16], end=[32])
+        out = proj_A + proj_B
+        return out
+
+    after = Rewriter(before)
+    tvm.ir.assert_structural_equal(expected, after)
+
```

Comment:
Can we add a test case for matching with "any" datatype?

Reply:
Can do.  I added a test case with a pattern-match against arbitrary datatype (last slice of a 2-d tensor), which can be applied to any known dtype, or an unknown dtype.